apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs

```diff
@@ -464,27 +464,29 @@ private void EvictAndRefreshSessions()
 
                 void RemoveStaleOrExpiredItemsFromStack(ConcurrentStack<PooledSession> stack)
                 {
-                    int count = stack.Count;
-                    for (int i = 0; i < count; i++)
+                    if (!stack.IsEmpty)
                     {
-                        // This shouldn't happen, as nothing else should touch the stack while we're in a
-                        // lock, but we don't want to throw an exception
-                        if (!stack.TryPop(out var session))
+                        int count = stack.Count;
+                        PooledSession[] sessions = new PooledSession[count];
+                        if (stack.TryPopRange(sessions) > 0)
```

Comment:
I think it would be better to have:

```csharp
int count;
PooledSession[] sessions = new PooledSession[stack.Count];
if ((count = stack.TryPopRange(sessions)) > 0)
{
}
```

That way it's clearer that we won't be trying to use an array element which wasn't actually popped.
Probably not a problem due to the locking, but might make it more robust if we change the locking strategy later.

Reply:
Yep, done.

Reply:
Actually, undoing this, `TryPopRange` fails when the stack is empty. It calls [this](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentstack-1.trypoprange?view=netframework-4.8#System_Collections_Concurrent_ConcurrentStack_1_TryPopRange__0___System_Int32_System_Int32_) overload internally that fails [because `startIndex` is the same as the array length](https://referencesource.microsoft.com/#mscorlib/system/Collections/Concurrent/ConcurrentStack.cs,0c76dfd482eb5d09)

Reply:
Right, so we still need to avoid doing anything if the stack is empty, but the previous code would have failed too.
My point is we should use the return value of TryPopRange to know how many items we really need to handle.

Reply:
Agreed, I'm doing a combination of both codes. Will flag when done.

Reply:
@jskeet PTAL, thanks!!